### DISCUSSION TITLE
Hard coded compilation flags

### DISCUSSION THREAD

**Discussion by mandresm** [FESOM/fesom2#336] [Ideas]
Hi all,

Trying to get some of the AWI coupled models ported into Levante I found that, for some of the components, it is not possible to define the compilation flags without editing the CMake files. One of those components is FESOM. These poses 2 problems:
- ESM-Tools can not pass to the model the compilation flags depending on the machine
- There are many branches of FESOM around, which means adding and `if` condition in the CMake files on all of these branches, to handle Levante's special requirements will be painful. That's what I did in this commit in a branch for AWICM3: e4d79462124d12b19bbef9880cbffa1e8963afec

That's a strategy copied from Kai Himstedt: If the environment variable `CPU_MODEL=AMD_EPYC_ZEN3` is set, then the `-march` and `-mtune` flags needed for Levante are added to the list of compilation flags. Otherwise, compilation flags remain a default.

Alternatively, we could do the same but instead of defining a `CPU_MODEL` defining directly `FESOM_FFLAGS`, and if this variable existS in the environment, then CMake should use that instead of the default compilation flags in FESOM.

In my view, no matter what we do, we are gonna have to change the `CMakeLists.txt` in a number of branches (same goes for ECHAM), but since we will have to do it anyway, I think it is now a good time to discuss what is the best strategy to follow.

@dsidoren @koldunovn @hegish @JanStreffing @pgierz

**Comment by koldunovn** [comment]
I am not at all CMake expert, so don't know what is the best solution, but can say that setting `-march=core-avx2 -mtune=core-avx2` on levante makes model run 2 times faster, so some solution is certainly needed :)
I would go for `FESOM_FFLAGS` as this is something easy to set in the `shell` file.

**Comment by hegish** [comment]
Technically it is not a problem to pass compiler flags to cmake fro the command line. See also here https://cmake.org/cmake/help/latest/envvar/FFLAGS.html?highlight=fflags
The problem with this approach is, this applies to every lib and exe we build from this cmake call. I think we had the compiler flags explicitly set in the CMakeLists.txt to have a well tested configuration. You can break many things with the wrong flags. So you can not just pass the same flags to everything we build.
@mandresm, what is it you want to change in the current setup? And for which cmake targets (i.e. libs, exe)?

**Reply by mandresm** [reply]
Thanks for the link to the documentation in CMake. I am no expert on CMake. Similarly to what @koldunovn mentioned above, these `-march=core-avx2 -mtune=core-avx2` flags are needed in Levante for running FESOM fast. In my commit I have modify the following line in `CMakeLists.txt` to that end (similarly to what @dsidoren has done):
```
target_compile_options(${PROJECT_NAME} PRIVATE -r8 -i4 -march=core-avx2 -mtune=core-avx2 -fp-model precise -no-prec-div -no-prec-sqrt -fast-transcendentals -ip -init=zero -no-wrap-margin)
```

My question to you would be what approach do you recommend to add these 2 flags for the compilation of FESOM, only in Levante?
- Some type of evaluation in the CMakeLists to catch if it's running in Levante as done here: https://github.com/FESOM/fesom2/commit/e4d79462124d12b19bbef9880cbffa1e8963afec
- pass compiler flags to cmake through ESM-Tools, when it finds that is running in Levante
- other solutions...

**Reply by koldunovn** [reply]
Can we have something like `FESOM_model` that will be passed only to the model compilation, and maybe `FESOM_partitioner` that will be passed for partitioner compilation, and some sane defaults if those environmental variables are not present?

**Comment by hegish** [comment]
If this is so critical, we should also add these flags to the other parts, e.g. the solver, right?
Is there a reason you add the flags to the intel compiler section? I thought levante is AMD?
Which compiler are you using on levante?

**Reply by mandresm** [reply]
I am using Intel because is what they recommend: https://docs.dkrz.de/doc/levante/code-development/compiling-and-linking.html#levante-compiling-and-linking

**Reply by hegish** [reply]
Ah OK, via oneapi. I forgot.
Would it help if we add the march and mtune flags if it is an AMD CPU? I assume this is nothing special for levante and we could use these flags on any AMD system where we choose tu use intel.

**Reply by mandresm** [reply]
Yes, that would certainly help. I just don't know how to best do that (specially taking into account that this is a change that will have to be applied for multiple branches of FESOM). Would someone from FESOM do it?

**Reply by koldunovn** [reply]
I did a PR #338 that can serve as a base. With this settings fesom compiles and run core experiments on levante.

**Reply by hegish** [reply]
There are no flags yet in the PR, right? I will try to do that tomorrow.

**Reply by mandresm** [reply]
Thanks to both!

**Reply by koldunovn** [reply]
No, no flags. The optimal (for now), would be

```
if(${CMAKE_Fortran_COMPILER_ID} STREQUAL  Intel )
	   target_compile_options(${PROJECT_NAME} PRIVATE -r8 -i4 -march=core-avx2 -mtune=core-avx2 -fp-model precise -no-prec-div -no-prec-sqrt -fimf-use-svml -
ip -init=zero -no-wrap-margin)
```

So I removed `-xHost` as it conflicts with `march`. Things might change, of course, and we might not only have to add this two flags fro levante.

**Comment by hegish** [comment]
Should I add my commits to the [levante_update](https://github.com/FESOM/fesom2/tree/levante_update) branch?

**Reply by koldunovn** [reply]
Yes, why not.

**Reply by hegish** [reply]
do you know if the -march=core-avx2 -mtune=core-avx2 flags are valid for the C compiler as well? I also want to build parms this way.

**Reply by mandresm** [reply]
I think so. Here (https://docs.dkrz.de/doc/levante/code-development/compiling-and-linking.html) in the "Intel Compilers" section states that `"The table below lists some useful compiler options that are commonly used for the Intel compilers."` where the table below include those flags. I understand that these would include C compiler.

**Reply by hegish** [reply]
> do you know if the -march=core-avx2 -mtune=core-avx2 flags are valid for the C compiler as well? I also want to build parms this way.

the flags seem to work for parms as well

**Reply by hegish** [reply]
I have updated https://github.com/FESOM/fesom2/tree/levante_update, let me know what you think.

**Reply by mandresm** [reply]
Thanks a lot @hegish . By looking at your changes I think I understand that `target_compile_options` can be used to expand the compiling flag list, right? This is something I was wondering how to do so I've learnt a bit :)

I just have one question, I thought you where going to make the additional flags be dependent on whether it was an AMD processor or not, but instead you are checking against the `FESOM_PLATFORM_STRATEGY` export variable. Would you clarify why in the end you chose to go for this strategy instead?

**Reply by hegish** [reply]
I could not automatically distinguish the CPU type from a typical Intel CPU. For example `uname -m` just says x86_64 on levante. Which is also the variable cmake sets for its CMAKE_HOST_SYSTEM_PROCESSOR variable. So we seem to have to explicitly tell cmake to do a special "levante" build.
To do this I utilized the env.sh script, where we already have each target platform hardcoded. No need to add another variable for each env script.
One can always overwrite this with e.g.
`cmake -DFESOM_PLATFORM_STRATEGY=obscure.platform ../src`

**Reply by mandresm** [reply]
Okay, thanks for the insight!

**Comment by hegish** [comment]
Should we use this approach for the other branches as well? @mandresm, could you provide me with a list of fesom branches you have to deal with?

**Reply by mandresm** [reply]
I think that would be great. 

@pgierz @JanStreffing @tsemmler05 please add to this if I am missing some branch:

Note: branches in GitHub unless specified otherwise.

**AWI-CM3 FESOM 2 branches**
- [x] `awicm-3-frontiers_parallel-restart` (the last commit here was by me trying to do what you did, you can revert it if you wish)
- [x] `awicm3-frontiers-freeze-candidate-5` (the last commit here was by me trying to do what you did, you can revert it if you wish)

**AWI-ESM-2 FESOM 2 branches**
- [ ] `2.1-awiesm-2.1`
- [ ] `wiso` if @mwerner-awi agrees
- [ ] `awiesm2-recom-wiso` (in https://gitlab.dkrz.de/FESOM/fesom2.git), maybe the interested parties should take care of that since it is in gitlab.dkrz (@mbutzin, @a270105)

**AWI-CM1 FESOM 1.4 branches**
- [x] `master` in https://gitlab.dkrz.de/modular_esm/fesom-1.4.git
- [x] `fesom_recom_1.4_master` in https://gitlab.dkrz.de/modular_esm/fesom-1.4.git if Judith Hauck and the REcoM team agree on that

**Reply by hegish** [reply]
someone should first check if fesom is now running at the desired speed on levante

**Reply by mandresm** [reply]
Hi @hegish,

I tested FESOM-2.0 cherry picking all the changes in #338 into a new branch `2.0_levante` and tested there. It seems it is improving the performance for levante (se metrics below). I also tested similar changes in `FESOM-1.4` also with improvements in the performance.

I'll take care of the mergings in the FESOM-1.4 branches. I have also pull-requested merges for FESOM-2.0 branches relevant for `AWICM3`.

My changes for `FESOM-2.0` that I use for the testing are in the `2.0_levante` branch which departed from the tag `2.0.2`. What should we do in terms of tagging/merging this new development for the `2.0.x` version?

**FESOM-2.0** (monthly run, 128 tasks, CORE2)
- Mistral (144 tasks): **3m6s**
- Without `-march` and `-mtune` flags: **9m17s**
- With `-march` and `-mtune` flags: **3m24s**

**FESOM-1.4** (mothly run, 384 tasks, CORE2)
- Mistral (288 tasks): **22m12s**
- Without `-march` and `-mtune` flags: **42m53s**
- With `-march` and `-mtune` flags: **9m59s**

**Reply by hegish** [reply]
> What should we do in terms of tagging/merging this new development for the `2.0.x` version?

Hm, @koldunovn, what do you say, do we have a version number pattern for this?

In any case @mandresm, please tell me if I can help with the changes/merges in the other branches.

**Reply by pgierz** [reply]
The AWI-ESM branches are all the ones that (I know of) we support. I will take care of the main one today, for recom and wiso @mbutzin and @mwerner-awi will need to have a look.

**Reply by pgierz** [reply]
OK, the main AWI-ESM changes are in https://github.com/FESOM/fesom2/pull/345. I will test it "in real life" as soon as Levante is back online if it compiles, but perhaps @hegish & @mandresm you can already have a look at what I would change.

**Reply by mwerner-awi** [reply]
> The AWI-ESM branches are all the ones that (I know of) we support. I will take care of the main one today, for recom and wiso @mbutzin and @mwerner-awi will need to have a look.

I don't understand all the technical details you are discussing here. But for the wiso part of AWIESM, no extra compiler flags are used. Everything is set via namelist entries.

**Reply by pgierz** [reply]
Thanks @mwerner-awi. I will then assume that if the standard AWI-ESM works, the wiso variant will work as well and then check off both boxes on @mandresm list once I can confirm that it compiles correctly.

**Reply by pgierz** [reply]
And additional info in case it went under in the techno-babble: this has to do with a potentially significant model speedup using optimised compiler flags suggested by DKRZ.
